test/tap/git-races.js

```diff
@@ -0,0 +1,171 @@
+})
+
+function cleanup () {
+  process.chdir(tmpdir())
+  rimraf.sync(wd)
+}
+
+var npm = requireInject.installGlobally('../../lib/npm.js', {
+  'child_process': {
+    'execFile': function (cmd, args, options, cb) {
+      // If it's a clone we swap any requests for any of the urls we're mocking
+      // with the path to the bare repo
+      if (args[0] === 'clone') {
+        var m2 = args.length - 2
+        var m1 = args.length - 1
+        if (testrepos[args[m2]]) {
+          testurls[args[m1]] = args[m2]
+          args[m2] = testrepos[args[m2]]
+        }
+        execFile(cmd, args, options, cb)
+      // here, we intercept npm validating the remote origin url on one of the
+      // clones we've done previously and return the original url that was requested
+      } else if (args[0] === 'config' && args[1] === '--get' && args[2] === 'remote.origin.url') {
+        process.nextTick(function () {
+          cb(null, testurls[options.cwd], '')
+        })
+      } else {
+        execFile(cmd, args, options, cb)
+      }
+    }
+  }
+})
+
+function extract (tarball, target, cb) {
+  cb = once(cb)
+  fs.createReadStream(tarball).on('error', function (er) { cb(er) })
+    .pipe(zlib.createGunzip()).on('error', function (er) { cb(er) })
+    .pipe(tar.Extract({path: target})).on('error', function (er) { cb(er) })
+    .on('end', function () {
+      cb()
+    })
+}
+
+// Copied from lib/utils/git, because we need to use
+// it before calling npm.load and lib/utils/git uses npm.js
+// which doesn't allow that. =( =(
+
+function prefixGitArgs () {
+  return process.platform === 'win32' ? ['-c', 'core.longpaths=true'] : []
+}
+
+var gitcmd
+
+function execGit (args, options, cb) {
+  var fullArgs = prefixGitArgs().concat(args || [])
+  return execFile(gitcmd, fullArgs, options, cb)
+}
+
+function gitWhichAndExec (args, options, cb) {
+  if (gitcmd) return execGit(args, options, cb)
+
+  which('git', function (err, pathtogit) {
+    if (err) {
+      err.code = 'ENOGIT'
+      return cb(err)
+    }
+    gitcmd = pathtogit
+
+    execGit(args, options, cb)
+  })
+}
+
+function andClone (gitdir, repodir, cb) {
+  return function (er) {
+    if (er) return cb(er)
+    gitWhichAndExec(['clone', gitdir, repodir], {}, cb)
+  }
+}
+
+function setup (cb) {
+  cleanup()
+  mkdirp.sync(wd)
+
+  extract(testcase_tgz, wd, andClone(testcase_git, testcase_path, andExtractPackages))
+
+  function andExtractPackages (er) {
+    if (er) return cb(er)
+    asyncMap(testtarballs, function (tgz, done) {
+      extract(tgz, wd, done)
+    }, andChdir)
+  }
+  function andChdir (er) {
+    if (er) return cb(er)
+    process.chdir(testcase_path)
+    andLoadNpm()
+  }
+  function andLoadNpm () {
+    var opts = {
+      cache: path.resolve(wd, 'cache')
+    }
+    npm.load(opts, cb)
+  }
+}
+
+test('setup', function (t) {
+  setup(function (er) {
+    t.ifError(er, 'setup ran OK')
+    t.end()
+  })
+})
+
+test('off-to-the-races', function (t) {
```

Comment:
Yeah, that's a good idea: I've added it w/ https://github.com/npm/npm/commit/cf703cbcaf61f6aee549c0bc819d4fe86824fddd